hard/gf_14/bsg_misc/bsg_mux_gen.py

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/python
+#
+# bsg_mux_gen_tsmc_40
+#
+# This script generates one hot muxes.
+#
+# data_i:   input data
+# data_o:   output data
+# sel_one_hot_i  :  one hot signal
+#           0 - recycle data
+#           1 - take data from previous node
+#           2 - take data from data_i
+#
+#
+# there are a few ways to implement the 3-input one hot mux:
+#
+# AOI222X1:     need an additional inverter. area 8.1 delay .17 to .33
+# NAND2, NAND3:      NAND2 delay:  .03 to .05; NAND3 delay: .04 to .11
+#
+# only real issue with using the NAND's is hold time...
+#
+
+import sys;
+
+def emit_module_header (name, input_args, output_args) :
+    print "module " + name + "(",
+    my_list = []
+    for x in input_args :
+        my_list.append("input "+x+"\n");
+    for x in output_args :
+        my_list.append("output "+x+"\n");
+    print (" "*(len(name)+8)+",").join(my_list);
+
+    print ");";
+
+def emit_module_footer( ) :
+    print "endmodule";
+
+def emit_wire_definition (name) :
+    print "wire " + name + "; "
+
+def emit_wire_definition_nocr (name) :
+    print "wire " + name + "; ",
+
+def emit_gate_instance (gate_str, arg_list ) :
+    print gate_instance(gate_str,arg_list);
+
+def queue_gate_instance (out_dict, gate_str, arg_list, order) :
+    the_string = gate_instance(gate_str,arg_list)
+    out_dict[the_string] = order
+
+def gate_instance (gate_str, arg_list ) :
+    for i in range(0,len(arg_list)) :
+        gate_str = gate_str.replace("#"+str(i),arg_list[i]);
+    return "// synopsys rp_orient ({N FS} " + arg_list[0] + ")\n" + gate_str;
+
+def access_bit (name, bit) :
+    return name + "[" + str(bit) + "]";
+
+def param_bits_all (name, bit) :
+    return "[" + str(bit-1) + ":0] " + name;
+
+def ident_name_word_bit (name,word,bit) :
+    return name + "_w" + str(word) + "_b" + str(bit);
+
+def ident_name_bit_port (name,bit,port) :
+    return name + "_b" + str(bit) + "_p" + str(port);
+
+def ident_name_word_bit_port (name,word,bit,port) :
+    return name + "_w" + str(word) + "_b" + str(bit) + "_p" + str(port);
+
+def ident_name_bit (name,bit) :
+    return name + "_b" + str(bit);
+
+
+def emit_rp_group_begin (name) :
+    print "// synopsys rp_group (" + name + ")"
+
+def emit_rp_group_end (name) :
+    print "// synopsys rp_endgroup (" + name +")"
+
+def emit_rp_fill (params):
+    print "// synopsys rp_fill (" + params +")"
+
+
+
+# NOTE: for symmetric pins, assume that earlier ones are always faster.
+# For example, for AOI22  A's are faster than B's and A0 is faster than A1.
+#
+
+fab = "gf_14"
+
+#nand2 = "ND2D2BWP #0 (.A1 (#1), .A2 (#2), .ZN(#3)                    );"
+#nand3 = "ND3D2BWP #0 (.A1 (#1), .A2 (#2), .A3 (#3), .ZN(#4)          );"
+#nand4 = "ND4D2BWP #0 (.A1 (#1), .A2 (#2), .A3 (#3), .A4 (#4), .ZN(#5));"
+
+nand2 = "SC7P5T_ND2X2_SSC14SL #0 (.A(#1), .B(#2), .Z(#3));"
```

Comment:
comment with justification (or lack of justification) for using high leakage cells